### ISSUE TITLE
[POC] SPDK

### ISSUE THREAD

**Issue by yasker** [longhorn/longhorn#760]
Maybe can replace tgtd.

Answer the question:
- [ ] Is it possible to replace tgtd to create iscsi target for a userspace process? And how long it will take? What's the performance/CPU utilization gain?
- [ ] How much effort it would need to create a kernel module to talk with SPDK to implement the block device interface?

**Comment by yasker** [comment]
Push to v0.7.1. SPDK requires huge page support which will definitely bring more complexity into containerizing it.

**Comment by yasker** [comment]
Kubernetes v1.14 introduced huge page management directly from container spec. We should able to utilize it to make it easier to run SPDK. https://kubernetes.io/docs/tasks/manage-hugepages/scheduling-hugepages/

**Comment by boknowswiki** [comment]
Initially adding longhorn module in spdk is done. And have done some basic testing read/write/dd command/fio testing.

**Comment by boknowswiki** [comment]
Initial test plan:
Since we need to run longhorn-engine as binary on my vm as a standalone process, and it will run as one replica and one engine two roles. And both rancher/tgtd and spdk need to run as standalone processes as well. To get the most similar testing environment, I will use taskset with cpu mask to make sure during tgtd or spdk test, all the necessary processes run on the same cpu say cpu 1 in my case, and using the same test script from [here](https://github.com/leeliu/dbench) and my  modifications are:
1. For the write test cases, I added "--end_fsync=1" by the end.
2. Since the space limitation, the block device created is 1GB, so the FIO_SIZE=500M and FIO_OFFSET_INCREMENT=200M.

Also will save the top output as references for cpu usage.

**Comment by boknowswiki** [comment]
Test report for tgtd with longhorn:

First confirmed all the processes are running on the same cpu:
`sudo taskset 0x2 ./bin/longhorn replica --listen localhost:9502 --size 1g ~/tmp/spdk/longhorn/lh_dir/

botao@botao_ubuntu:~/tmp/spdk/longhorn$ ps -ea f | grep longhorn
  3889 pts/0    S+     0:00  |           \_ grep --color=auto longhorn
  3875 pts/4    S+     0:00  |           \_ sudo taskset 0x2 ./bin/longhorn replica --listen localhost:9502 --size 1g /home/botao/tmp/spdk/longhorn/lh_dir/
  3876 pts/4    Sl+    0:00  |               \_ ./bin/longhorn replica --listen localhost:9502 --size 1g /home/botao/tmp/spdk/longhorn/lh_dir/
  3883 pts/4    Sl+    0:00  |                   \_ /home/botao/go/src/github.com/longhorn/longhorn-engine/bin/longhorn sync-agent --listen localhost:9504 --replica localhost:9502 --listen-port-range 9505-9514
botao@botao_ubuntu:~/tmp/spdk/longhorn$ ps -p 3876 -L -o pid,ppid,tid,psr,pcpu
   PID   PPID    TID PSR %CPU
  3876   3875   3876   1  0.0
  3876   3875   3877   1  0.0
  3876   3875   3878   1  0.0
  3876   3875   3879   1  0.0
  3876   3875   3880   1  0.0
  3876   3875   3881   1  0.0
  3876   3875   3882   1  0.0


sudo taskset 0x2 ./usr/tgtd -f

botao@botao_ubuntu:~/tmp/spdk/longhorn$ ps -ea f | grep tgtd
  3926 pts/0    S+     0:00  |           \_ grep --color=auto tgtd
  3923 pts/7    S+     0:00              \_ sudo taskset 0x2 ./usr/tgtd -f
  3924 pts/7    S+     0:00                  \_ ./usr/tgtd -f
botao@botao_ubuntu:~/tmp/spdk/longhorn$ ps -p 3924 -L -o pid,ppid,tid,psr,pcpu
   PID   PPID    TID PSR %CPU
  3924   3923   3924   1  0.0


sudo taskset 0x2 ./bin/longhorn controller --replica tcp://127.0.0.1:9502 test --frontend tgt-blockdev

botao@botao_ubuntu:~/tmp/spdk/longhorn$ ps -ea f | grep longhorn
  4046 pts/0    S+     0:00  |           \_ grep --color=auto longhorn
  3930 pts/3    S+     0:00  |           \_ sudo taskset 0x2 ./bin/longhorn controller --replica tcp://127.0.0.1:9502 test --frontend tgt-blockdev
  3931 pts/3    Sl+    0:00  |               \_ ./bin/longhorn controller --replica tcp://127.0.0.1:9502 test --frontend tgt-blockdev
  3875 pts/4    S+     0:00  |           \_ sudo taskset 0x2 ./bin/longhorn replica --listen localhost:9502 --size 1g /home/botao/tmp/spdk/longhorn/lh_dir/
  3876 pts/4    Sl+    0:00  |               \_ ./bin/longhorn replica --listen localhost:9502 --size 1g /home/botao/tmp/spdk/longhorn/lh_dir/
  3883 pts/4    Sl+    0:00  |                   \_ /home/botao/go/src/github.com/longhorn/longhorn-engine/bin/longhorn sync-agent --listen localhost:9504 --replica localhost:9502 --listen-port-range 9505-9514
botao@botao_ubuntu:~/tmp/spdk/longhorn$ ps -p 3931 -L -o pid,ppid,tid,psr,pcpu
   PID   PPID    TID PSR %CPU
  3931   3930   3931   1  0.0
  3931   3930   3932   1  0.0
  3931   3930   3933   1  0.0
  3931   3930   3934   1  0.0
  3931   3930   3935   1  0.0
  3931   3930   3936   1  0.0
  3931   3930   3937   1  0.0
  3931   3930   3949   1  0.0
`
After got `/dev/sdb`, format to ext4 and mount it to a tmp directory.
Run `docker-entrypoint.sh` and write the output to fio_x.log x is from 1 to 3. Meaning run the test 3 times.
Below are the 3 times summary:
`fio_1.log:= Dbench Summary =
fio_1.log-==================
fio_1.log-Random Read/Write IOPS: 3493/2873. BW: 198MiB/s / 177MiB/s
fio_1.log-Average Latency (usec) Read/Write: 1092.03/1409.70
fio_1.log-Sequential Read/Write: 280MiB/s / 309MiB/s
fio_1.log-Mixed Random Read/Write IOPS: 2463/824
--
fio_2.log:= Dbench Summary =
fio_2.log-==================
fio_2.log-Random Read/Write IOPS: 3385/2689. BW: 196MiB/s / 179MiB/s
fio_2.log-Average Latency (usec) Read/Write: 1064.59/1407.90
fio_2.log-Sequential Read/Write: 277MiB/s / 311MiB/s
fio_2.log-Mixed Random Read/Write IOPS: 2480/834
--
fio_3.log:= Dbench Summary =
fio_3.log-==================
fio_3.log-Random Read/Write IOPS: 3528/2874. BW: 194MiB/s / 183MiB/s
fio_3.log-Average Latency (usec) Read/Write: 1061.84/1380.70
fio_3.log-Sequential Read/Write: 280MiB/s / 310MiB/s
fio_3.log-Mixed Random Read/Write IOPS: 2466/828`

Avg cpu usage for tgtd is around 29%

**Comment by boknowswiki** [comment]
Test report for spdk longhorn module:

First confirmed all the processes are running on the same cpu
`
Start longhorn replica:

sudo taskset 0x2 ./bin/longhorn replica --listen localhost:9502 --size 1g ~/tmp/spdk/longhorn/lh_dir/

$ ps -ea f | grep longhorn
  4758 pts/0    S+     0:00  |           \_ grep --color=auto longhorn
  4743 pts/4    S+     0:00  |           \_ sudo taskset 0x2 ./bin/longhorn replica --listen localhost:9502 --size 1g /home/botao/tmp/spdk/longhorn/lh_dir/
  4744 pts/4    Sl+    0:00  |               \_ ./bin/longhorn replica --listen localhost:9502 --size 1g /home/botao/tmp/spdk/longhorn/lh_dir/
  4749 pts/4    Sl+    0:00  |                   \_ /home/botao/go/src/github.com/longhorn/longhorn-engine/bin/longhorn sync-agent --listen localhost:9504 --replica localhost:9502 --listen-port-range 9505-9514
$ ps -p 4749 -L -o pid,ppid,tid,psr,pcpu
   PID   PPID    TID PSR %CPU
  4749   4744   4749   1  0.0
  4749   4744   4751   1  0.0
  4749   4744   4752   1  0.0
  4749   4744   4753   1  0.0
  4749   4744   4755   1  0.0
  4749   4744   4756   1  0.0

Start longhorn-engine:

sudo taskset 0x2 ./bin/longhorn controller --replica tcp://127.0.0.1:9502 volume-test --frontend socket

$ ps -ea f | grep longhorn
  4772 pts/0    S+     0:00  |           \_ grep --color=auto longhorn
  4761 pts/3    S+     0:00  |           \_ sudo taskset 0x2 ./bin/longhorn controller --replica tcp://127.0.0.1:9502 volume-test --frontend socket
  4762 pts/3    Sl+    0:00  |               \_ ./bin/longhorn controller --replica tcp://127.0.0.1:9502 volume-test --frontend socket
  4743 pts/4    S+     0:00  |           \_ sudo taskset 0x2 ./bin/longhorn replica --listen localhost:9502 --size 1g /home/botao/tmp/spdk/longhorn/lh_dir/
  4744 pts/4    Sl+    0:00  |               \_ ./bin/longhorn replica --listen localhost:9502 --size 1g /home/botao/tmp/spdk/longhorn/lh_dir/
  4749 pts/4    Sl+    0:00  |                   \_ /home/botao/go/src/github.com/longhorn/longhorn-engine/bin/longhorn sync-agent --listen localhost:9504 --replica localhost:9502 --listen-port-range 9505-9514
$ps -p 4762 -L -o pid,ppid,tid,psr,pcpu
   PID   PPID    TID PSR %CPU
  4762   4761   4762   1  0.0
  4762   4761   4763   1  0.0
  4762   4761   4764   1  0.0
  4762   4761   4765   1  0.0
  4762   4761   4766   1  0.0
  4762   4761   4767   1  0.0
  4762   4761   4768   1  0.0

Start spdk longhorn module:

sudo taskset 0x2 ./build/bin/iscsi_tgt -m 0x2

$ ps -ea f | grep tgt
  4855 pts/0    S+     0:00  |           \_ grep --color=auto tgt
  4851 pts/1    S+     0:00  |           \_ sudo taskset 0x2 ./build/bin/iscsi_tgt -m 0x2
  4852 pts/1    RLl+   0:04  |               \_ ./build/bin/iscsi_tgt -m 0x2
$ ps -p 4852 -L -o pid,ppid,tid,psr,pcpu
   PID   PPID    TID PSR %CPU
  4852   4851   4852   1 97.0
  4852   4851   4853   1  0.0
`
After got `/dev/sdb`, format to ext4 and mount it to a tmp directory.
Run `docker-entrypoint.sh` and write the output to fio_x.log x is from 1 to 3. Meaning run the test 3 times.
Below are the 3 times summary:

`fio_1.log:= Dbench Summary =
fio_1.log-==================
fio_1.log-Random Read/Write IOPS: 3842/2937. BW: 97.8MiB/s / 85.6MiB/s
fio_1.log-Average Latency (usec) Read/Write: 1154.68/1486.91
fio_1.log-Sequential Read/Write: 110MiB/s / 111MiB/s
fio_1.log-Mixed Random Read/Write IOPS: 2563/851

fio_2.log:= Dbench Summary =
fio_2.log-==================
fio_2.log-Random Read/Write IOPS: 3785/3021. BW: 93.6MiB/s / 85.2MiB/s
fio_2.log-Average Latency (usec) Read/Write: 1112.08/1416.71
fio_2.log-Sequential Read/Write: 111MiB/s / 110MiB/s
fio_2.log-Mixed Random Read/Write IOPS: 2566/864

fio_3.log:= Dbench Summary =
fio_3.log-==================
fio_3.log-Random Read/Write IOPS: 3787/3016. BW: 99.7MiB/s / 87.2MiB/s
fio_3.log-Average Latency (usec) Read/Write: 1126.83/1468.82
fio_3.log-Sequential Read/Write: 112MiB/s / 113MiB/s
fio_3.log-Mixed Random Read/Write IOPS: 2724/903`

`Avg cpu usage for spdk is around 24%`
`Avg mem usage for spdk is around %24%`

**Comment by yasker** [comment]
IOPS is higher with SPDK but bandwidth is lower? That's interesting. We need to look into it more later.

**Comment by boknowswiki** [comment]
One thing noticed during the testing, there is one time the replica connection got dropped, put the log here:

Got time out on longhorn-engine:
```
ERRO[0481] Read timeout on replcia 127.0.0.1:9503 seq= 556262 size= 64 (kB) 
ERRO[0486] Ping timeout on replica 127.0.0.1:9503 seq= 556263 
ERRO[0486] Retry  1 on replica 127.0.0.1:9503 seq= 556263 size= 0 (kB) 
ERRO[0486] Retry  1 on replica 127.0.0.1:9503 seq= 556262 size= 64 (kB) 
ERRO[0482] Read timeout on replcia 127.0.0.1:9503 seq= 556260 size= 64 (kB) 
ERRO[0487] Retry  1 on replica 127.0.0.1:9503 seq= 556260 size= 64 (kB) 
ERRO[0482] Read timeout on replcia 127.0.0.1:9503 seq= 556261 size= 64 (kB) 
ERRO[0483] Read timeout on replcia 127.0.0.1:9503 seq= 556249 size= 64 (kB) 
ERRO[0489] Retry  1 on replica 127.0.0.1:9503 seq= 556249 size= 64 (kB) 
ERRO[0484] Read timeout on replcia 127.0.0.1:9503 seq= 556257 size= 64 (kB) 
ERRO[0489] Retry  1 on replica 127.0.0.1:9503 seq= 556257 size= 64 (kB) 
ERRO[0485] Read timeout on replcia 127.0.0.1:9503 seq= 556255 size= 64 (kB) 
ERRO[0489] Retry  1 on replica 127.0.0.1:9503 seq= 556255 size= 64 (kB) 
ERRO[0485] Read timeout on replcia 127.0.0.1:9503 seq= 556256 size= 64 (kB) 
ERRO[0489] Retry  1 on replica 127.0.0.1:9503 seq= 556256 size= 64 (kB) 
ERRO[0485] Read timeout on replcia 127.0.0.1:9503 seq= 556252 size= 64 (kB) 
ERRO[0489] Retry  1 on replica 127.0.0.1:9503 seq= 556252 size= 64 (kB) 
ERRO[0485] Read timeout on replcia 127.0.0.1:9503 seq= 556254 size= 64 (kB) 
ERRO[0489] Retry  1 on replica 127.0.0.1:9503 seq= 556254 size= 64 (kB) 
ERRO[0485] Read timeout on replcia 127.0.0.1:9503 seq= 556258 size= 64 (kB) 
ERRO[0489] Retry  1 on replica 127.0.0.1:9503 seq= 556258 size= 64 (kB) 
ERRO[0486] Read timeout on replcia 127.0.0.1:9503 seq= 556253 size= 64 (kB) 
ERRO[0490] Retry  1 on replica 127.0.0.1:9503 seq= 556253 size= 64 (kB) 
ERRO[0486] Read timeout on replcia 127.0.0.1:9503 seq= 556250 size= 64 (kB) 
ERRO[0490] Retry  1 on replica 127.0.0.1:9503 seq= 556250 size= 64 (kB) 
ERRO[0486] Read timeout on replcia 127.0.0.1:9503 seq= 556259 size= 64 (kB) 
ERRO[0490] Retry  1 on replica 127.0.0.1:9503 seq= 556259 size= 64 (kB) 
ERRO[0489] Retry  1 on replica 127.0.0.1:9503 seq= 556261 size= 64 (kB)
```
And spdk connection dropped, spdk longhorn module message:
```
[2020-07-02 14:36:31.149007] conn.c: 951:iscsi_drop_conns: *ERROR*: drop old connections iqn.2016-06.io.spdk:disk1 by iqn.1993-08.org.debian:01:62ebda3b76cc,i,0x00023d000001
[2020-07-02 14:36:31.163870] conn.c: 955:iscsi_drop_conns: *ERROR*: exiting conn by iqn.1993-08.org.debian:01:62ebda3b76cc,i,0x00023d000001 (127.0.0.1)
[2020-07-02 14:36:31.634717] conn.c: 974:iscsi_drop_conns: *ERROR*: exiting 1 conns
[2020-07-02 14:36:33.669706] conn.c: 951:iscsi_drop_conns: *ERROR*: drop old connections iqn.2016-06.io.spdk:disk1 by iqn.1993-08.org.debian:01:62ebda3b76cc,i,0x00023d000001
[2020-07-02 14:36:33.686452] conn.c: 955:iscsi_drop_conns: *ERROR*: exiting conn by iqn.1993-08.org.debian:01:62ebda3b76cc,i,0x00023d000001 (127.0.0.1)
[2020-07-02 14:36:33.957874] conn.c: 974:iscsi_drop_conns: *ERROR*: exiting 1 conns
```
spdk kernel message:
```
[  590.885223]  connection1:0: ping timeout of 5 secs expired, recv timeout 5, last rx 4295037610, last ping 4295038864, now 4295040128
[  590.885278]  connection1:0: detected conn error (1022)
[  595.218014]  connection1:0: Could not send nopout
[  595.218095]  connection1:0: detected conn error (1020)
[  599.162769] sd 33:0:0:0: [sdb] tag#0 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  599.162771] sd 33:0:0:0: [sdb] tag#0 Sense Key : Aborted Command [current] 
[  599.162772] sd 33:0:0:0: [sdb] tag#0 Add. Sense: No additional sense information
[  599.162773] sd 33:0:0:0: [sdb] tag#0 CDB: Read(10) 28 00 00 1c b7 00 00 01 00 00
[  599.162774] print_req_error: I/O error, dev sdb, sector 1881856
[  599.162835] sd 33:0:0:0: [sdb] tag#1 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  599.162836] sd 33:0:0:0: [sdb] tag#1 Sense Key : Aborted Command [current] 
[  599.162837] sd 33:0:0:0: [sdb] tag#1 Add. Sense: No additional sense information
[  599.162838] sd 33:0:0:0: [sdb] tag#1 CDB: Read(10) 28 00 00 1c 8e 00 00 01 00 00
[  599.162838] print_req_error: I/O error, dev sdb, sector 1871360
[  599.163249] sd 33:0:0:0: [sdb] tag#2 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  599.163249] sd 33:0:0:0: [sdb] tag#2 Sense Key : Aborted Command [current] 
[  599.163250] sd 33:0:0:0: [sdb] tag#2 Add. Sense: No additional sense information
[  599.163251] sd 33:0:0:0: [sdb] tag#2 CDB: Read(10) 28 00 00 1c a5 00 00 01 00 00
[  599.163251] print_req_error: I/O error, dev sdb, sector 1877248
[  599.163637] sd 33:0:0:0: [sdb] tag#3 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  599.163637] sd 33:0:0:0: [sdb] tag#3 Sense Key : Aborted Command [current] 
[  599.163638] sd 33:0:0:0: [sdb] tag#3 Add. Sense: No additional sense information
[  599.163638] sd 33:0:0:0: [sdb] tag#3 CDB: Read(10) 28 00 00 1c a7 00 00 01 00 00
[  599.163639] print_req_error: I/O error, dev sdb, sector 1877760
[  599.164105] sd 33:0:0:0: [sdb] tag#4 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  599.164106] sd 33:0:0:0: [sdb] tag#4 Sense Key : Aborted Command [current] 
[  599.164107] sd 33:0:0:0: [sdb] tag#4 Add. Sense: No additional sense information
[  599.164108] sd 33:0:0:0: [sdb] tag#4 CDB: Read(10) 28 00 00 1c c9 00 00 01 00 00
[  599.164108] print_req_error: I/O error, dev sdb, sector 1886464
[  599.164518] sd 33:0:0:0: [sdb] tag#5 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  599.164518] sd 33:0:0:0: [sdb] tag#5 Sense Key : Aborted Command [current] 
[  599.164519] sd 33:0:0:0: [sdb] tag#5 Add. Sense: No additional sense information
[  599.164520] sd 33:0:0:0: [sdb] tag#5 CDB: Read(10) 28 00 00 1d 5a 00 00 01 00 00
[  599.164520] print_req_error: I/O error, dev sdb, sector 1923584
[  599.164894] sd 33:0:0:0: [sdb] tag#6 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  599.164895] sd 33:0:0:0: [sdb] tag#6 Sense Key : Aborted Command [current] 
[  599.164895] sd 33:0:0:0: [sdb] tag#6 Add. Sense: No additional sense information
[  599.164896] sd 33:0:0:0: [sdb] tag#6 CDB: Read(10) 28 00 00 1c f6 00 00 01 00 00
[  599.164897] print_req_error: I/O error, dev sdb, sector 1897984
[  599.165284] sd 33:0:0:0: [sdb] tag#7 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  599.165285] sd 33:0:0:0: [sdb] tag#7 Sense Key : Aborted Command [current] 
[  599.165285] sd 33:0:0:0: [sdb] tag#7 Add. Sense: No additional sense information
[  599.165286] sd 33:0:0:0: [sdb] tag#7 CDB: Read(10) 28 00 00 1d 0f 00 00 01 00 00
[  599.165287] print_req_error: I/O error, dev sdb, sector 1904384
[  599.165745] sd 33:0:0:0: [sdb] tag#8 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  599.165746] sd 33:0:0:0: [sdb] tag#8 Sense Key : Aborted Command [current] 
[  599.165747] sd 33:0:0:0: [sdb] tag#8 Add. Sense: No additional sense information
[  599.165748] sd 33:0:0:0: [sdb] tag#8 CDB: Read(10) 28 00 00 1d 41 00 00 01 00 00
[  599.165748] print_req_error: I/O error, dev sdb, sector 1917184
[  599.166117] sd 33:0:0:0: [sdb] tag#9 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  599.166118] sd 33:0:0:0: [sdb] tag#9 Sense Key : Aborted Command [current] 
[  599.166119] sd 33:0:0:0: [sdb] tag#9 Add. Sense: No additional sense information
[  599.166119] sd 33:0:0:0: [sdb] tag#9 CDB: Read(10) 28 00 00 1b 68 00 00 01 00 00
[  599.166120] print_req_error: I/O error, dev sdb, sector 1796096
[  602.641307] Aborting journal on device sdb-8.
[  602.894662] Buffer I/O error on dev sdb, logical block 131072, lost sync page write
[  602.895033] JBD2: Error -5 detected when updating journal superblock for sdb-8.
```

**Comment by yasker** [comment]
That's probably the reason why it's slower. We need to fix this and test again later.

**Comment by boknowswiki** [comment]
Summary spdk seems faster on IOPS but slower on bandwidth. And spdk seems takes lots of memory and cpu. During the spdk testing, my vm become noticeable slow, even only spdk running standalone, noticed the reactor from spdk is taking 100% cpu.

**Comment by boknowswiki** [comment]
Couple things we might take a further look at current liblonghorn:
1. We use `write_full` to write information to the socket, will they be packed together or will be sent separately? If separately, then there will be lots of small packages meaning lots of receiving.
2. Maybe we can put the time stamp inside the package and verify it after received it, in this case we can avoid one timeout handler thread to interact with the receiving thread with locking.

**Comment by boknowswiki** [comment]
> That's probably the reason why it's slower. We need to fix this and test again later after v1.0.1.

This time, the test is aborted. The 3 times tests are running without any errors.
